src/MongoDB.Driver.Core/Core/Compression/Zstd/Zstandard64NativeMethods.cs

```diff
@@ -0,0 +1,270 @@
+        {
+            var result = __ZSTD_endStream.Value(zcs, outputBuffer);
+            ThrowIfError(result);
+            return result;
+        }
+
+        public static ulong ZSTD_freeCStream(IntPtr zcs)
+        {
+            return __ZSTD_freeCStream.Value(zcs);
+        }
+
+        public static ulong ZSTD_freeDStream(IntPtr zds)
+        {
+            return __ZSTD_freeDStream.Value(zds);
+        }
+
+        public static ulong ZSTD_initDStream(IntPtr zds)
+        {
+            return __ZSTD_initDStream.Value(zds);
+        }
+
+        public static ulong ZSTD_decompressStream(IntPtr zds, [MarshalAs(UnmanagedType.LPStruct)] NativeBufferInfo outputBuffer, [MarshalAs(UnmanagedType.LPStruct)] NativeBufferInfo inputBuffer)
+        {
+            var result = __ZSTD_decompressStream.Value(zds, outputBuffer, inputBuffer);
+            ThrowIfError(result);
+            return result;
+        }
+
+        public static ulong ZSTD_initCStream(IntPtr zcs, int compressionLevel)
+        {
+            var result = __ZSTD_initCStream.Value(zcs, compressionLevel);
+            ThrowIfError(result);
+            return result;
+        }
+
+        public static ulong ZSTD_compressStream(IntPtr zcs, [MarshalAs(UnmanagedType.LPStruct)] NativeBufferInfo outputBuffer, [MarshalAs(UnmanagedType.LPStruct)] NativeBufferInfo inputBuffer)
+        {
+            var result = __ZSTD_compressStream.Value(zcs, outputBuffer, inputBuffer);
+            ThrowIfError(result);
+            return result;
+        }
+
+        // private static methods
+        private static Lazy<TDelegate> CreateLazyForDelegate<TDelegate>(string name)
+        {
+            return new Lazy<TDelegate>(() => __libraryLoader.Value.GetDelegate<TDelegate>(name), isThreadSafe: true);
+        }
+
+        private static void ThrowIfError(ulong code)
+        {
+            if (Zstandard64NativeMethods.ZSTD_isError(code))
+            {
+                var errorPtr = Zstandard64NativeMethods.ZSTD_getErrorName(code);
+                var errorMsg = Marshal.PtrToStringAnsi(errorPtr);
+                throw new IOException(errorMsg);
+            }
+        }
+
+        private static bool ZSTD_isError(ulong code)
+        {
+            return __ZSTD_isError.Value(code);
+        }
+
+        private static IntPtr ZSTD_getErrorName(ulong code)
+        {
+            return __ZSTD_getErrorName.Value(code);
+        }
+
+        // nested types
+        private class Delegates64
+        {
+            public delegate ulong ZSTD_CStreamInSize();
+            public delegate ulong ZSTD_CStreamOutSize();
+            public delegate IntPtr ZSTD_createCStream();
+
+            public delegate ulong ZSTD_DStreamInSize();
+            public delegate ulong ZSTD_DStreamOutSize();
+            public delegate IntPtr ZSTD_createDStream();
+
+            public delegate int ZSTD_maxCLevel();
```

Comment:
I suspect this might actually return `long`.

Reply:
You misunderstood. It's not about the max value you expect to be returned.

It's about matching the declaration in the C library (in zstd.h):

    ZSTDLIB_API int         ZSTD_maxCLevel(void);

An `int` in C corresponds to a `long` in C# (for 64-bit libraries).